File: 2204-find-subsequence-of-length-k-with-the-largest-sum/find-subsequence-of-length-k-with-the-largest-sum.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer[]}
def max_subsequence(nums, k)
    # Mỗi phần tử là [giá trị, chỉ số]
    indexed = nums.each_with_index.map { |val, i| [val, i] }

    # Lấy k phần tử có giá trị lớn nhất (theo val)
    top_k = indexed.sort_by { |val, _| -val }[0...k]

    # Sắp xếp lại theo chỉ số ban đầu để giữ thứ tự subsequence
    sorted = top_k.sort_by { |_, i| i }

    # Trả về mảng kết quả
    sorted.map { |val, _| val }
end

File: 2204-find-subsequence-of-length-k-with-the-largest-sum/find-subsequence-of-length-k-with-the-largest-sum.go
type Pair struct {
	val int
	idx int
}

type MaxHeap []Pair

func (h MaxHeap) Len() int            { return len(h) }
func (h MaxHeap) Less(i, j int) bool  { return h[i].val > h[j].val } // Max heap
func (h MaxHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(Pair)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	val := old[n-1]
	*h = old[:n-1]
	return val
}

func maxSubsequence(nums []int, k int) []int {
	h := &MaxHeap{}
	for i, num := range nums {
		heap.Push(h, Pair{num, i})
	}

	topK := make([]Pair, k)
	for i := 0; i < k; i++ {
		topK[i] = heap.Pop(h).(Pair)
	}

	// Sắp xếp lại theo chỉ số để giữ thứ tự subsequence
	sort.Slice(topK, func(i, j int) bool {
		return topK[i].idx < topK[j].idx
	})

	result := make([]int, k)
	for i := 0; i < k; i++ {
		result[i] = topK[i].val
	}
	return result
}